spikeextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -1145,45 +1161,62 @@ def write_units(
                     spike_train_frames = sorting.get_unit_spike_train(unit_id=unit_id)
-                    assert spike_train_frames[-1] < len(timestamps), "Number of 'timestamps' differs from number of " \
-                                                                     "'frames'!"
-                    spkt = np.array(timestamps)[spike_train_frames]
+                    assert spike_train_frames[-1] < len(timestamps), \
+                        "Number of 'timestamps' differs from number of 'frames'!"
+                    unit_kwargs.update(spike_times=np.array(timestamps)[spike_train_frames])
                 else:
-                    spkt = sorting.get_unit_spike_train(unit_id=unit_id) / fs
+                    unit_kwargs.update(spike_times=sorting.get_unit_spike_train(unit_id=unit_id) / sf)
+
                 for pr in write_properties:
                     if pr in sorting.get_unit_property_names(unit_id):
                         prop_value = sorting.get_unit_property(unit_id, pr)
                         unit_kwargs.update({pr: prop_value})
                     else:  # Case of missing data for this unit and this property
                         unit_kwargs.update({pr: np.nan})
-                nwbfile.add_unit(id=unit_id, spike_times=spkt, **unit_kwargs)
 
-            # TODO
-            # # Stores average and std of spike traces
-            # This will soon be updated to the current NWB standard
-            # if 'waveforms' in sorting.get_unit_spike_feature_names(unit_id=id):
-            #     wf = sorting.get_unit_spike_features(unit_id=id,
-            #                                          feature_name='waveforms')
-            #     relevant_ch = most_relevant_ch(wf)
-            #     # Spike traces on the most relevant channel
-            #     traces = wf[:, relevant_ch, :]
-            #     traces_avg = np.mean(traces, axis=0)
-            #     traces_std = np.std(traces, axis=0)
-            #     nwbfile.add_unit(
-            #         id=id,
-            #         spike_times=spkt,
-            #         waveform_mean=traces_avg,
-            #         waveform_sd=traces_std
-            #     )
+                if write_waveforms:
+                    check_waveform_features(sorting=sorting, unit_id=unit_id)
+                    wf = sorting.get_unit_spike_features(unit_id=unit_id, feature_name='waveforms')
+                    # This is temporary until the NWB Schema allows conversion factors for waveforms in the units table
+                    wf = wf * gain * 1e-6
+                    # ---
+                    unit_kwargs.update(waveforms=wf)
+
+                if write_waveform_stats:
+                    check_waveform_features(sorting=sorting, unit_id=unit_id)
+                    wf = sorting.get_unit_spike_features(unit_id=unit_id, feature_name='waveforms')
+                    relevant_ch = most_relevant_ch(wf)
+                    traces = wf[:, :, relevant_ch]
```

Comment:
waveforms in SI have the following dimensions: (num_spikes, num_channels, num_samples)
```suggestion
                    traces = wf[:, relevant_ch]
```

Reply:
Fixed